### PR TITLE
Added tests for all operations that were added that are supported by Mbed Crypto

### PR DESCRIPTION
**Pull request by sbailey-arm**
Tests added for:
`psa_copy_key`
`psa_aead_encrypt`
`psa_aead_decrypt`
`psa_hash_compute`
`psa_hash_compare`
`psa_raw_key_agreement`

Signed-off-by: Samuel Bailey <samuel.bailey@arm.com>

### CODE REVIEW

File: psa-crypto/tests/key_agreement.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+const EXPECTED_OUTPUT: [u8; 32] = [
+    0xd6, 0x84, 0x0f, 0x6b, 0x42, 0xf6, 0xed, 0xaf, 0xd1, 0x31, 0x16, 0xe0, 0xe1, 0x25, 0x65, 0x20,
+    0x2f, 0xef, 0x8e, 0x9e, 0xce, 0x7d, 0xce, 0x03, 0x81, 0x24, 0x64, 0xd0, 0x4b, 0x94, 0x42, 0xde,
+];
+
+#[test]
+fn key_agreement() {
+    let alg = RawKeyAgreement::Ecdh;
+    let attributes = Attributes {
+        key_type: Type::EccKeyPair {
```
Comment (hug-dev) [inline psa-crypto/tests/key_agreement.rs:28]: So key agreement can also be done on ECC keys? Maybe we need to update [this code](https://docs.rs/psa-crypto/0.3.0/src/psa_crypto/types/key.rs.html#262) then.
Is it the case that `Ecdh` can be used with ECC keys where `Ffdh` can be used with DH keys?

File: psa-crypto/Cargo.toml [hunk 2]
```diff
@@ -25,7 +25,7 @@ rand = "0.7.3"
 base64 = "0.12.3"
 
 [features]
-default = ["operations", "no-std"]
+default = ["operations"]
```
[the remark this answers is not in the document]
Comment (sbailey-arm) [reply psa-crypto/Cargo.toml:28]: 😨 sorry, that slipped through the net, it should definitely not be removed. I removed it for debugging (for `dbg!`)

File: psa-crypto/src/operations/key_management.rs [hunk 3]
```diff
@@ -298,7 +298,9 @@ pub(crate) fn complete_new_key_operation(
     handle: psa_key_handle_t,
 ) -> Result<Id> {
     if key_lifetime != Lifetime::Volatile {
-        Status::from(unsafe { psa_crypto_sys::psa_close_key(handle) }).to_result()?;
+        Status::from(unsafe { psa_crypto_sys::psa_close_key(handle) })
+            .to_result()
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (sbailey-arm) [reply psa-crypto/src/operations/key_management.rs:303]: Yeah, this was another debug thing. Seems I did not check the diff like I normally do 😬.

File: psa-crypto/tests/key_agreement.rs (hunk 1 above)
Comment (sbailey-arm) [reply psa-crypto/tests/key_agreement.rs:28]: Yes, ECC supports key agreement. Not sure about second question, maybe @ionut-arm has an idea? It also [doesn't look like Mbed Crypto supports FFDH at the moment](https://github.com/ARMmbed/mbedtls/issues/3261#issue-607690687).

Comment (ionut-arm) [reply psa-crypto/tests/key_agreement.rs:28]: Yes and yes. I actually asked a similar question back in the day and I think they've included clarification on this in subsequent releases of the spec - found it weird that they didn't explicitly specify that ECC keys are used for `ECDH` when they have a DH key type.

The "classical" Diffie-Hellman (over finite fields) uses DH keys and is represented by `FFDH`, the elliptic-curve variant uses ECC keys and is represented by `ECDH`.

Comment (ionut-arm) [reply psa-crypto/tests/key_agreement.rs:28]: Yeah, that code seems wrong, good spot!

Comment (hug-dev) [reply psa-crypto/tests/key_agreement.rs:28]: Cool, thanks for the clarification! In that case, could you update it as part of this PR @sbailey-arm please 😃 ?

File: psa-crypto/src/types/key.rs [hunk 4]
```diff
@@ -259,15 +259,23 @@ impl Attributes {
...
-                }
-            }
+            Type::EccKeyPair { .. } | Type::EccPublicKey { .. } => match alg {
+                Algorithm::KeyAgreement(KeyAgreement::Raw(RawKeyAgreement::Ecdh))
+                | Algorithm::KeyAgreement(KeyAgreement::WithKeyDerivation {
+                    ka_alg: RawKeyAgreement::Ecdh,
+                    ..
+                }) => true,
+                Algorithm::AsymmetricSignature(sign_alg) => sign_alg.is_ecc_alg(),
+                Algorithm::AsymmetricEncryption(_) => true,
```
Comment (hug-dev) [inline psa-crypto/src/types/key.rs:269]: Hmm I think I told you something wrong, there don't seem to be any [asymmetric encryption algorithms](https://parallaxsecond.github.io/parsec-book/parsec_client/operations/psa_algorithm.html#asymmetricencryption-algorithm) for ECC keys.

File: psa-crypto/src/operations/key_management.rs [hunk 5]
```diff
@@ -322,29 +322,28 @@ pub(crate) fn complete_new_key_operation(
...
...0x02, 0x41, 0x00, 0xb6, 0xcd, 0xbd, 0x35, 0x4f, 0x7d, 0xf5, 0x79, 0xa6, 0x3b, 0x48, 0xb3, 0x64, 0x3e, 0x35, 0x3b, 0x84, 0x89, 0x87, 0x77, 0xb4, 0x8b, 0x15, 0xf9, 0x4e, 0x0b, 0xfc, 0x05, 0x67, 0xa6, 0xae, 0x59, 0x11, 0xd5, 0x7a, 0xd6, 0x40, 0x9c, 0xf7, 0x64, 0x7b, 0xf9, 0x62, 0x64, 0xe9, 0xbd, 0x87, 0xeb, 0x95, 0xe2, 0x63, 0xb7, 0x11, 0x0b, 0x9a, 0x1f, 0x9f, 0x94, 0xac, 0xce, 0xd0, 0xfa, 0xfa, 0x4d, 0x02, 0x40, 0x71, 0x19, 0x5e, 0xec, 0x37, 0xe8, 0xd2, 0x57, 0xde, 0xcf, 0xc6, 0x72, 0xb0, 0x7a, 0xe6, 0x39, 0xf1, 0x0c, 0xbb, 0x9b, 0x0c, 0x73, 0x9d, 0x0c, 0x80, 0x99, 0x68, 0xd6, 0x44, 0xa9, 0x4e, 0x3f, 0xd6, 0xed, 0x92, 0x87, 0x07, 0x7a, 0x14, 0x58, 0x3f, 0x37, 0x90, 0x58, 0xf7, 0x6a, 0x8a, 0xec, 0xd4, 0x3c, 0x62, 0xdc, 0x8c, 0x0f, 0x41, 0x76, 0x66, 0x50, 0xd7, 0x25, 0x27, 0x5a, 0xc4, 0xa1, 0x02, 0x41, 0x00, 0xbb, 0x32, 0xd1, 0x33, 0xed, 0xc2, 0xe0, 0x48, 0xd4, 0x63, 0x38, 0x8b, 0x7b, 0xe9, 0xcb, 0x4b, 0xe2, 0x9f, 0x4b, 0x62, 0x50, 0xbe, 0x60, 0x3e, 0x70, 0xe3, 0x64, 0x75, 0x01, 0xc9, 0x7d, 0xdd, 0xe2, 0x0a, 0x4e, 0x71, 0xbe, 0x95, 0xfd, 0x5e, 0x71, 0x78, 0x4e, 0x25, 0xac, 0xa4, 0xba, 0xf2, 0x5b, 0xe5, 0x73, 0x8a, 0xae, 0x59, 0xbb, 0xfe, 0x1c, 0x99, 0x77, 0x81, 0x44, 0x7a, 0x2b, 0x24];
```
[the remark this answers is not in the document]
Comment (sbailey-arm) [reply psa-crypto/src/operations/key_management.rs:326]: Nope, that looks pretty horrific! I've hashed it out.

[the remark this answers is not in the document]
Comment (ionut-arm) [reply psa-crypto/src/operations/key_management.rs]: 👌  Sometimes it's actually useful to have some values that work out of the box for testing, but they can just look at the code.

**Pull request merged**